### PR TITLE
Only offer encapsulate on fields of types

### LINKED ISSUE #282: ArgumentException at AbstractCodeGenerationService.CheckDeclarationNode
Ported from TFS WorkItem: <b>1096007</b>

------------------------------

<p><b>Repro Steps:</b></p>
<P><B>Affected Branch:</B> <BR><B>Affected Build:</B> <BR><B>Affected Product Language:</B> <BR><BR><B>Steps to reproduce:</B><BR><BR>From activity log:</P>
<P>&nbsp;</P>
<P>&lt;description&gt;System.AggregateException: One or more errors occurred. ---&amp;gt; System.ArgumentException: Destination type must be a EnumDeclarationSyntax, TypeDeclarationSyntax or CompilationUnitSyntax.<BR>Parameter name: destination<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.CheckDeclarationNode[TDeclarationNode1,TDeclarationNode2,TDeclarationNode3](SyntaxNode destination)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddField[TDeclarationNode](TDeclarationNode destination, IFieldSymbol field, CodeGenerationOptions options, IList`1 availableIndices)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.&amp;lt;&amp;gt;c__DisplayClass35.&amp;lt;AddFieldAsync&amp;gt;b__36(SyntaxNode t, CodeGenerationOptions opts, IList`1 ai)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.&amp;lt;GetEditAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.GenerateMember.GenerateVariable.AbstractGenerateVariableService`3.GenerateVariableCodeAction.&amp;lt;GetChangedDocumentAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;GetChangedSolutionAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;ComputeOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;ComputePreviewOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;GetPreviewOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.&amp;lt;&amp;gt;c__DisplayClass8.&amp;lt;&amp;lt;GetPreview&amp;gt;b__10&amp;gt;d__0.MoveNext()<BR>&nbsp;&nbsp; --- End of inner exception stack trace ---<BR>&nbsp;&nbsp; at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)<BR>&nbsp;&nbsp; at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.&amp;lt;&amp;gt;c__DisplayClass8.&amp;lt;GetPreview&amp;gt;b__9()<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunction[T](IExtensionManager extensionManager, Object extension, Func`1 function, T defaultValue)<BR>---&amp;gt; (Inner Exception #0) System.ArgumentException: Destination type must be a EnumDeclarationSyntax, TypeDeclarationSyntax or CompilationUnitSyntax.<BR>Parameter name: destination<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.CheckDeclarationNode[TDeclarationNode1,TDeclarationNode2,TDeclarationNode3](SyntaxNode destination)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddField[TDeclarationNode](TDeclarationNode destination, IFieldSymbol field, CodeGenerationOptions options, IList`1 availableIndices)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.&amp;lt;&amp;gt;c__DisplayClass35.&amp;lt;AddFieldAsync&amp;gt;b__36(SyntaxNode t, CodeGenerationOptions opts, IList`1 ai)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.&amp;lt;GetEditAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.GenerateMember.GenerateVariable.AbstractGenerateVariableService`3.GenerateVariableCodeAction.&amp;lt;GetChangedDocumentAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;GetChangedSolutionAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;ComputeOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;ComputePreviewOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.CodeActions.CodeAction.&amp;lt;GetPreviewOperationsAsync&amp;gt;d__1.MoveNext()<BR>--- End of stack trace from previous location where exception was thrown ---<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)<BR>&nbsp;&nbsp; at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)<BR>&nbsp;&nbsp; at Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.&amp;lt;&amp;gt;c__DisplayClass8.&amp;lt;&amp;lt;GetPreview&amp;gt;b__10&amp;gt;d__0.MoveNext()&amp;lt;---<BR>&lt;/description&gt;<BR></P>

------------------------------

<p><b>Revisions:</b></p>

1) Created By David Poeschl (12/14/2014 3:34:17 PM)

Who owns the code generation service?

------------------------------

2) Edited By Kevin Pilch-Bisson (12/15/2014 9:44:47 AM)

Will this go away with Matt's changes?

------------------------------

3) Edited By Srivatsn Narayanan (12/15/2014 10:27:48 AM)

When Matt moves the IDE over to the new API and deletes codegenerationservice, it would go away i suppose :)

------------------------------

### PR DESCRIPTION
**Pull request by mattwar**
Fixes #282 

This change limits the Encapsulate Field code fix to only offer fixes for fields that are members of a type.

@Pilchie @jasonmalinowski  @dpoeschl  @rchande  please review

### CODE REVIEW

**Comment by rchande** [comment]
Tests? Otherwise :+1:

**Comment by jasonmalinowski** [comment]
Indeed, please add tests.

File: src/Features/CSharp/EncapsulateField/CSharpEncapsulateFieldService.cs [hunk 1]
```diff
@@ -130,6 +130,12 @@ internal class CSharpEncapsulateFieldService : AbstractEncapsulateFieldService
                                 .Where(f => f.Name.Length != 0);
         }
 
+        private bool CanEncapsulate(FieldDeclarationSyntax field)
+        {
+            return field.Parent is TypeDeclarationSyntax
+                || field.Parent is EnumDeclarationSyntax;
```
Comment (dpoeschl) [inline src/Features/CSharp/EncapsulateField/CSharpEncapsulateFieldService.cs]: What does encapsulate field do on enums? An EnumDeclarationSyntax can't contain a FieldDeclarationSyntax, right?

Comment (mattwar) [inline src/Features/CSharp/EncapsulateField/CSharpEncapsulateFieldService.cs]: You are right.  I was concerned with only fixing the problem that it not cause the code generator to assert/throw, so I basically copied the constraint from the code gen.  But really, this feature cannot work on enums either, so I'll remove that.

**Comment by Pilchie** [comment]
:+1:

**Comment by mattwar** [comment]
@dotnet-bot please test this

**Comment by mattwar** [comment]
@dotnet-bot please test this

**Comment by mattwar** [comment]
@dotnet-bot please test this

**Comment by mattwar** [comment]
@dotnet-bot please test this

**Comment by dpoeschl** [comment]
retest this please

**Pull request merged**